### PR TITLE
#1340, Closes #1180: Prepare Quickstart for Drupal core 9.3.x.

### LINKED ISSUE #1180: Drupal 9.3.x compatibility
## Motivation
Drupal 9.3.0 will be released Dec 8, 2021. 

## Proposed Resolution
We need to prepare Quickstart for compatibility with this release.

## Additional context
https://www.drupal.org/project/drupal/releases/9.3.0-alpha1
https://www.drupal.org/project/drupal/releases/9.3.0-beta1
https://www.drupal.org/project/drupal/releases/9.3.0-beta2
https://www.drupal.org/project/drupal/releases/9.3.0-beta3
https://www.drupal.org/project/drupal/releases/9.3.0-rc1

### PR DESCRIPTION
**Pull request by joeparsons**
## Description
- Re-exported ALL configuration included with Quickstart to apply sorting changes from Drupal core
  - See https://www.drupal.org/node/3230199
- Updated various core patches (still need to upload final versions of patches in use to Drupal.org)
- Removed some permissions from authenticated and anonymous roles to prevent config dependencies on unused modules (e.g. comment and contact)
  - See https://www.drupal.org/node/3193348 
- Addressed various 9.3.x code deprecations:
  - https://www.drupal.org/node/2940438
  - https://www.drupal.org/node/2940031

### Remaining tasks:
- [x] Switch from 9.3.0-rc1 to 9.3.0 once released (before merging)
- [x] Finalize updated patches (replace drupal.org MR diffs with actual patches)
- [ ] Add DB update for uninstalling quickedit (?)
  - See https://www.drupal.org/node/3227039 
- ~Add DB update for installing access_unpublished (?)~
- [x] Check if optional HTML paragraph module changes user role permissions (and potentially re-export)
- [ ] Peer review & testing!

## Related Issue

- Closes #1180
- #1340 

## How Has This Been Tested?

Locally with lando

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I've added this Pull Request to the AZ Quickstart Github project and set it to "Review in progress".
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_core/config/install/core.entity_view_mode.node.az_row.yml
Comment (trackleft) [inline modules/custom/az_core/config/install/core.entity_view_mode.node.az_row.yml:1]:
```suggestion
```

File: modules/custom/az_google_tag/config/install/google_tag.container.az_comprehensive.yml
Comment (trackleft) [inline modules/custom/az_google_tag/config/install/google_tag.container.az_comprehensive.yml:45]: I wonder what is happening here

File: composer.json
Comment (trackleft) [inline composer.json:55]:
```suggestion
        "drupal/core-recommended": "9.3.3",
```

**Comment by trackleft** [comment]
This should be ready for you now @tadean

**Comment by trackleft** [comment]
FYI one of the [core patches](https://git.drupalcode.org/project/drupal/-/merge_requests/1489.diff) no longer applies after 9.3.3, starting with 9.3.4 [core/modules/media/tests/fixtures/update/drupal-9.2.0.standard.media-installed.php.gz](https://git.drupalcode.org/project/drupal/-/merge_requests/1489/diffs#diff-content-6ae11092ffe6a5f6f3e66f33b358b3d7153acad0)
seems must have changed.

File: composer.json
Comment (danahertzberg) [inline composer.json:55]: Consider changing to version 9.3.12

**Comment by joeparsons** [comment]
> FYI one of the [core patches](https://git.drupalcode.org/project/drupal/-/merge_requests/1489.diff) no longer applies after 9.3.3, starting with 9.3.4 [core/modules/media/tests/fixtures/update/drupal-9.2.0.standard.media-installed.php.gz](https://git.drupalcode.org/project/drupal/-/merge_requests/1489/diffs#diff-content-6ae11092ffe6a5f6f3e66f33b358b3d7153acad0) seems must have changed.

I've rebased the [MR on the drupal.org issue](https://www.drupal.org/project/drupal/issues/3109289#comment-14502007)

File: az_quickstart.info.yml
Comment (joeparsons) [inline az_quickstart.info.yml:78]: This shouldn't be needed with the changes from Troy's PR.
```suggestion
```

File: composer.json
Comment (joeparsons) [inline composer.json:131]: Replace MR with patch
```suggestion
                "Unnecessary restrictions on logo format: Can't upload replacement SVG logo (2259567)": "https://www.drupal.org/files/issues/2022-05-05/2259567-160-9-3-x.patch",
```

File: composer.json
Comment (joeparsons) [inline composer.json:134]: Use updated d.o. patch:
```suggestion
                "Allow text field to enforce a specific text format (784672)": "https://www.drupal.org/files/issues/2022-03-27/784672-227.patch",
```

**Review by mmunro-ltrr — APPROVED** [review]
image
Merged in Troy's PR, merged in main, looks OK apart from these log messages.

File: modules/custom/az_news/config/install/core.entity_view_display.node.az_news.az_marquee.yml
Comment (bberndt-uaz) [inline modules/custom/az_news/config/install/core.entity_view_display.node.az_news.az_marquee.yml:53]:
```suggestion
        classes: 'border-bottom align-text-top mb-4'
```

File: modules/custom/az_news/config/install/core.entity_view_display.node.az_news.default.yml
Comment (bberndt-uaz) [inline modules/custom/az_news/config/install/core.entity_view_display.node.az_news.default.yml:58]:
```suggestion
        classes: 'border-bottom align-text-top mb-4'
```

**Comment by joeparsons** [comment]
The PHP notices that were showing up in the PHPunit logs were coming from `liuggio/fastest` (see https://github.com/liuggio/fastest/issues/176).  We need to update to the latest version in our dev metapackage (i.e. merge az-digital/az-quickstart-dev#39) before merging this.

We should also update phpstan-drupal before merging this (see az-digital/az-quickstart-dev#38).

File: modules/custom/az_event/config/install/core.entity_view_display.node.az_event.az_card.yml
Comment (bberndt-uaz) [inline modules/custom/az_event/config/install/core.entity_view_display.node.az_event.az_card.yml:178]:
```suggestion
        classes: 'card card-borderless card-clickable bg-gray-100 mt-0 mb-4 p-0'
```

File: modules/custom/az_news/config/install/core.entity_view_display.node.az_news.az_marquee.yml
Comment (bberndt-uaz) [inline modules/custom/az_news/config/install/core.entity_view_display.node.az_news.az_marquee.yml:116]:
```suggestion
      label: Lead
```

File: modules/custom/az_news/config/install/core.entity_view_display.node.az_news.default.yml
Comment (bberndt-uaz) [inline modules/custom/az_news/config/install/core.entity_view_display.node.az_news.default.yml:127]:
```suggestion
      label: Lead
```

**Comment by joeparsons** [comment]
Final install / test time comparisons vs main branch.  These are acceptable increases IMHO.

It's worth noting that another PR that was recently merged into main added an additional test (#1400)

<table>
<tr>
	<td>Task
	<td>main (#1565)
	<td>This PR (latest revision)
<tr>
	<td>Install Quickstart
	<td>1 minutes 36 seconds 523 milliseconds
	<td>2 minutes 17 seconds 885 milliseconds
<tr>
	<td>PHPunit tests
	<td>12 minutes 32 seconds 854 milliseconds
	<td>16 minutes 29 seconds 798 milliseconds
<tr>
	<td>Migration tests
	<td>4 minutes 2 seconds 373 milliseconds
	<td>4 minutes 29 seconds 990 milliseconds
</table>

**Review by trackleft — APPROVED** [review]

**Comment by trackleft** [comment]
> image Merged in Troy's PR, merged in main, looks OK apart from these log messages.

@mmunro-ltrr Added a new issue for this: https://github.com/az-digital/az_quickstart/issues/1564

**Comment by joeparsons** [comment]
@trackleft @tadean I just rebuilt this PR locally in lando, did a fresh Quickstart install and full re-export of all the config again.  I _think_ this is ready for final review now but I also think we're probably going to need to manually restore some of the permissions that are being removed.

**Comment by joeparsons** [comment]
@tadean Not exactly sure what's going on with some of the permissions getting removed when I re-export roles.  I even tried installing more optional modules (comment, contact, webform, etc) before doing so but some permissions still seem to be getting removed...

**Comment by joeparsons** [comment]
Maybe we just need to revert the changes to the exported roles in this branch before merging for now.

**Comment by tadean** [comment]
Taking a look at this...

**Comment by tadean** [comment]
Okay, i think I understand what's going on. It's expected that doing an export is removing permissions since they aren't active, but we'd hoped to handle that in a follow up (some convenience method that merges active permissions with stored file permissions to make it easy to add more without having to enable every single optional module in the export.

However, even with every optional module something odd is still happening. My list that I enabled was
```
az_news_feeds
access_unpublished
antibot
comments
contact
webform
webform_access
webform_node
book
content_translation
az_carousel
environment_indicator
quick_node_clone
```
 And that seemed to cover most of the permissions. However, one type of permissions are still being cut out:
 ```
 create terms in tags
 delete any book content
 delete own webform content
 ```
 
Notice that these are permissions that are generated by other modules, but they are actually owned by the Node module. In our permission trimming code, we only only enable permissions from **newly enabled** modules. This is to prevent Quickstart from continually trying to reapply permissions a user had customized - it only changes permissions for modules that have been newly enabled:
```
  public function findProfilePermissions(array $extensions = []) {

    // Build list of permissions by providers (eg. module)
    $permissions_definitions = $this->getPermissionDefinitions();
    $permissions_by_provider = [];
    foreach ($permissions_definitions as $key => $permission) {
      $permissions_by_provider[$permission['provider']][] = $key;
    }
    // Filter list only to newly active permissions.
    $permissions_by_provider = array_intersect_key($permissions_by_provider, $extensions);
```
However, we're in a circumstance here where the presence of a module causes permission causes some permissions to be generated, but the module doesn't actually own the permissions, so they're skipped by our 'Only new permissions' code. In other words, if we were to somehow "re-enable" the node/taxonomy modules those permissions would probably be applied as normal. (Since the presence of Webform/Book/Etc. causes the permissions to be autogenerated, but Node is the actual owner and so forth).

Ideally we don't want to constantly renable all permissions old or new (because of the mentioned issue of messing with a user's customizations) so maybe what's wanted is a way of finding out if a permission is autogenerated based on a module or something so it can understand that those need to be turned on...

This may not be an issue directly since we don't specifically need to export the user roles again, but the issue does need to be solved since the permissions of autogenerated content types won't be correct. I'm looking into this...

**Comment by tadean** [comment]
Here's a permission definition for one of these permissions:
```
[delete any book content] =>; Array
    (
        [title] =&gt; Drupal\Core\StringTranslation\TranslatableMarkup Object
        [dependencies] => Array
            (
                [config] => Array
                    (
                        [0] => node.type.book
                    )
            )
        [description] =>
        [provider] => node
    )
 ```
 Note that this permission comes into existence when Book is enabled, but it really is owned by Node, thus causing our safeguard check to only affect permissions from newly enabled modules to ignore it. However, note that it depends on node.type.book. (a config entity that comes from the book module). I wonder if this might be a way to check for these 'Generated permissions'

**Comment by joeparsons** [comment]
Thanks for looking into this @tadean!  Do you think for the sake of getting this merged so we can create a release with 9.3.x that it would be best to just revert the changes to the `user.role.*.yml` files in this PR for now and address this as a follow-up?

**Comment by joeparsons** [comment]
After talking to @tadean @trackleft @danahertzberg and others, I think we've decided it would be best to merge this as-is after reverting the changes to the role files and then address the missing `Edit own/any X entity` issue as a follow-up so that we can proceed with preparing an initial release candidate.

I'll push a change that reverts the roles ASAP.

**Review by bberndt-uaz — APPROVED** [review]

**Pull request merged**
